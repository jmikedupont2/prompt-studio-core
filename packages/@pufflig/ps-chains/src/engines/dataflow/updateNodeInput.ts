import { ParamValue } from "@pufflig/ps-types";
import { Chain, NodeState, RunOptions } from "../../types";
import { applyDefaultInputs } from "./utils/utils";
import _ from "lodash";

/**
 * This function updates the input state of a node without executing the chain.
 *
 * @param nodeId id of the node to run
 * @param input input to update the node inputs with, empty object if existing inputs should be used
 * @param chain current definition and state of the chain
 * @param runOptions
 */
export async function updateNodeInput(
  nodeId: string,
  input: Record<string, ParamValue>,
  chain: Chain,
  runOptions?: RunOptions
) {
  const chainState: Record<string, NodeState> = { ...chain.state };
  const nodeConfig = chain.definition.nodes[nodeId];
  const nodeDefinition = chain.nodeTypes[nodeConfig?.type];

  if (!nodeConfig || !nodeDefinition) {
    throw new Error(`Definition for node ${nodeId} not found`);
  }

  // parse the input
  const nodeState = chainState[nodeId];
  const prevInput = nodeState?.input;
  const parsedInput = await nodeDefinition.parseInput(input, prevInput);

  // compile the new state of the node
  const newInput = { ...applyDefaultInputs(prevInput, nodeDefinition), ...parsedInput };
  const newState: NodeState = { ...nodeState, input: newInput };

  runOptions?.onNodeInputUpdate?.(nodeId, newState);

  const newChainState = _.merge(chainState, { [nodeId]: newState });
  return newChainState;
}
