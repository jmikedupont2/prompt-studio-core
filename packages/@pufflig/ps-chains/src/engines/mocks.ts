import { Chain } from "../types";

export const singleNode: Chain = {
  definition: {
    edges: [],
    nodes: [
      {
        id: "1",
        type: "input/template_editor",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    ],
  },
  state: {
    "1": {
      data: {},
      status: "idle",
    },
  },
};

export const simpleChain: Chain = {
  definition: {
    edges: [
      {
        id: "1",
        source: "1",
        target: "2",
        source_handle: "template",
        target_handle: "template",
      },
      {
        id: "2",
        source: "2",
        target: "3",
        source_handle: "text",
        target_handle: "prompt",
      },
    ],
    nodes: [
      {
        id: "1",
        type: "input/template_editor",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      {
        id: "2",
        type: "modifier/handlebar_template_completion",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      {
        id: "3",
        type: "output/completion_display",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    ],
  },
  state: {
    "1": {
      data: {},
      status: "idle",
    },
    "2": {
      data: {
        variables: [
          {
            id: "World",
            name: "World",
            description: "",
            type: "text",
            defaultValue: "mars",
          },
        ],
      },
      status: "idle",
    },
    "3": {
      data: {},
      status: "idle",
    },
  },
};

export const autorunExample: Chain = {
  definition: {
    edges: [
      {
        id: "1",
        source: "1",
        target: "2",
        source_handle: "template",
        target_handle: "template",
      },
      {
        id: "2",
        source: "2",
        target: "3",
        source_handle: "text",
        target_handle: "prompt",
      },
    ],
    nodes: [
      {
        id: "1",
        type: "input/template_editor",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      {
        id: "2",
        type: "modifier/handlebar_template_completion",
        autorun: false,
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      {
        id: "3",
        type: "output/completion_display",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    ],
  },
  state: {
    "1": {
      data: {},

      status: "idle",
    },
    "2": {
      data: {
        variables: [
          {
            id: "World",
            name: "World",
            description: "",
            type: "text",
            defaultValue: "mars",
          },
        ],
      },
      status: "idle",
    },
    "3": {
      data: {},
      status: "idle",
    },
  },
};

export const autorunRootExample: Chain = {
  definition: {
    edges: [
      {
        id: "1",
        source: "1",
        target: "2",
        source_handle: "template",
        target_handle: "template",
      },
      {
        id: "2",
        source: "2",
        target: "3",
        source_handle: "text",
        target_handle: "prompt",
      },
    ],
    nodes: [
      {
        id: "1",
        type: "input/template_editor",
        autorun: false,
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      {
        id: "2",
        type: "modifier/handlebar_template_completion",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      {
        id: "3",
        type: "output/completion_display",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    ],
  },
  state: {
    "1": {
      data: {},
      status: "idle",
    },
    "2": {
      data: {
        variables: [
          {
            id: "World",
            name: "World",
            description: "",
            type: "text",
            defaultValue: "mars",
          },
        ],
      },
      status: "idle",
    },
    "3": {
      data: {},
      status: "idle",
    },
  },
};

export const simpleLoop: Chain = {
  definition: {
    edges: [
      {
        id: "1",
        source: "1",
        target: "1",
        source_handle: "template",
        target_handle: "template",
      },
    ],
    nodes: [
      {
        id: "1",
        type: "input/template_editor",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    ],
  },
  state: {
    "1": {
      data: {},
      status: "idle",
    },
  },
};

export const missingStates: Chain = {
  definition: {
    nodes: [
      {
        id: "c96eeaef-a4bb-4fa6-a32f-b9ad176456ea",
        type: "adapter/openai_completion",
        autorun: false,
        editor: { position: { x: 0, y: 0 } },
      },
      {
        id: "197375bb-c777-4be5-a423-6d5618e2200f",
        type: "input/template_editor",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      {
        id: "85c8bc66-cc5a-4400-b4d3-bc119ed1b2f7",
        type: "modifier/handlebar_template_completion",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      {
        id: "2708636f-633c-4c02-abec-093fdc79f4d5",
        type: "output/completion_display",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
    ],
    edges: [
      {
        id: "5b253916-3ed9-4755-9f76-71adc451601f",
        source: "197375bb-c777-4be5-a423-6d5618e2200f",
        target: "85c8bc66-cc5a-4400-b4d3-bc119ed1b2f7",
        source_handle: "template",
        target_handle: "template",
      },
      {
        id: "2ad80e03-1ddd-4379-ae48-4a6398d20ff8",
        source: "85c8bc66-cc5a-4400-b4d3-bc119ed1b2f7",
        target: "c96eeaef-a4bb-4fa6-a32f-b9ad176456ea",
        source_handle: "text",
        target_handle: "prompt",
      },
      {
        id: "5ec1d394-6d14-47e0-a8ad-0b3b94371009",
        source: "85c8bc66-cc5a-4400-b4d3-bc119ed1b2f7",
        target: "2708636f-633c-4c02-abec-093fdc79f4d5",
        source_handle: "text",
        target_handle: "prompt",
      },
      {
        id: "5e2d0ac0-1e13-4313-a71d-d7bbcb522220",
        source: "c96eeaef-a4bb-4fa6-a32f-b9ad176456ea",
        target: "2708636f-633c-4c02-abec-093fdc79f4d5",
        source_handle: "completion",
        target_handle: "completion",
      },
    ],
  },
  state: {
    "197375bb-c777-4be5-a423-6d5618e2200f": {
      status: "idle",
      data: { template: "This is a test!" },
    },
  },
};

export const multiInput: Chain = {
  definition: {
    nodes: [
      {
        id: "1",
        type: "input/template_editor",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      {
        id: "2",
        type: "modifier/handlebar_template_completion",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      {
        id: "3",
        type: "output/completion_display",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
    ],
    edges: [
      {
        id: "5b253916-3ed9-4755-9f76-71adc451601f",
        source: "1",
        target: "2",
        source_handle: "template",
        target_handle: "template",
      },
      {
        id: "2ad80e03-1ddd-4379-ae48-4a6398d20ff8",
        source: "1",
        target: "3",
        source_handle: "template",
        target_handle: "prompt",
      },
      {
        id: "5ec1d394-6d14-47e0-a8ad-0b3b94371009",
        source: "2",
        target: "3",
        source_handle: "text",
        target_handle: "completion",
      },
    ],
  },
  state: {
    "1": {
      status: "idle",
      data: { template: "This is a test!" },
    },
  },
};
