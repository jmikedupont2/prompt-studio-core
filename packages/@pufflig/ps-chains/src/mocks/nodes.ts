import { Node, NodeActions } from "@pufflig/ps-types";

export const passthroughNode: NodeActions = {
  execute: async (i) => i,
  mapInput: async (i) => i,
};

export const loopNode: NodeActions = {
  getTargets: async ({ list }) => {
    return [
      { execSource: "exec:output", inputs: { data: list?.[0] } },
      { execSource: "exec:output", inputs: { data: list?.[1] } },
      { execSource: "exec:complete", inputs: {} },
    ];
  },
};

export const joinNode: NodeActions<{ data: string; list: string[] }> = {
  mapInput: async ({ data }, options) => {
    return { list: [...(options?.prevInput?.list || []), data] };
  },
};

export const simpleDataNode: Node = {
  name: "simpleNode",
  inputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  outputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  ...passthroughNode,
};

export const configOnlyNode: Node = {
  name: "simpleNode",
  inputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  outputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
};

export const multiInputDataNode: Node = {
  name: "multiInputNode",
  inputs: [
    {
      name: "data1",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data1",
    },
    {
      name: "data2",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data2",
    },
  ],
  outputs: [
    {
      name: "data1",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data1",
    },
  ],
  ...passthroughNode,
};

export const multiOutputDataNode: Node = {
  name: "multiOutputNode",
  inputs: [
    {
      name: "input1",
      type: "text",
      defaultValue: "",
      description: "",
      id: "input1",
    },
  ],
  outputs: [
    {
      name: "output1",
      type: "text",
      defaultValue: "",
      description: "",
      id: "output1",
    },
    {
      name: "output2",
      type: "text",
      defaultValue: "",
      description: "",
      id: "output2",
    },
  ],
  ...passthroughNode,
};

export const startNode: Node = {
  name: "startNode",
  execution: {
    inputs: [],
    outputs: [
      {
        id: "exec:output",
        name: "Completed",
      },
    ],
  },
  inputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  outputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  ...passthroughNode,
};

export const simpleExecNode: Node = {
  name: "simpleNode",
  execution: {
    inputs: [
      {
        id: "exec:input",
      },
    ],
    outputs: [
      {
        id: "exec:output",
        name: "Completed",
      },
    ],
  },
  inputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  outputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  ...passthroughNode,
};

export const loopNodeConfig: Node = {
  name: "loopNode",
  execution: {
    inputs: [
      {
        id: "exec:input",
      },
    ],
    outputs: [
      {
        id: "exec:output",
      },
    ],
  },
  inputs: [
    {
      name: "list",
      type: "list",
      defaultValue: [],
      description: "",
      id: "list",
    },
  ],
  outputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  ...loopNode,
};

export const joinNodeConfig: Node = {
  name: "joinNode",
  execution: {
    inputs: [
      {
        id: "exec:input",
      },
    ],
    outputs: [
      {
        id: "exec:output",
      },
    ],
  },
  inputs: [
    {
      name: "data",
      type: "text",
      defaultValue: "",
      description: "",
      id: "data",
    },
  ],
  outputs: [
    {
      name: "list",
      type: "list",
      defaultValue: [],
      description: "",
      id: "list",
    },
  ],
  ...joinNode,
};
