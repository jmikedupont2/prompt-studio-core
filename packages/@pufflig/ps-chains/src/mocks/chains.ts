import { Chain } from "../types";
import { asyncNode, multiInputNode, simpleNode, templateNode } from "./nodes";

export const singleNodeChain: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
  },
  definition: {
    edges: {},
    nodes: {
      "1": {
        id: "1",
        type: "simple_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    },
  },
  state: {
    "1": {
      input: {},
      status: "idle",
    },
  },
};

export const simpleChain: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
    template_node: templateNode,
  },
  definition: {
    edges: {
      e1: {
        id: "e1",
        source: "n1",
        target: "n2",
        sourceHandle: "template",
        targetHandle: "template",
      },
      e2: {
        id: "e2",
        source: "n2",
        target: "n3",
        sourceHandle: "text",
        targetHandle: "prompt",
      },
    },
    nodes: {
      n1: {
        id: "n1",
        type: "simple_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      n2: {
        id: "n2",
        type: "template_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      n3: {
        id: "n3",
        type: "simple_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    },
  },
  state: {
    n1: {
      input: {},
      status: "idle",
    },
    n2: {
      input: {
        variables: [
          {
            id: "World",
            name: "World",
            description: "",
            type: "text",
            defaultValue: "mars",
          },
        ],
      },
      status: "idle",
    },
    n3: {
      input: {},
      status: "idle",
    },
  },
};

export const autorunExample: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
    template_node: templateNode,
  },
  definition: {
    edges: {
      e1: {
        id: "e1",
        source: "n1",
        target: "n2",
        sourceHandle: "template",
        targetHandle: "template",
      },
      e2: {
        id: "e2",
        source: "n2",
        target: "n3",
        sourceHandle: "text",
        targetHandle: "prompt",
      },
    },
    nodes: {
      n1: {
        id: "n1",
        type: "simple_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      n2: {
        id: "n2",
        type: "template_node",
        autorun: false,
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      n3: {
        id: "n3",
        type: "simple_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    },
  },
  state: {
    n1: {
      input: {},

      status: "idle",
    },
    n2: {
      input: {
        variables: [
          {
            id: "World",
            name: "World",
            description: "",
            type: "text",
            defaultValue: "mars",
          },
        ],
      },
      status: "idle",
    },
    n3: {
      input: {},
      status: "idle",
    },
  },
};

export const autorunRootExample: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
    template_node: templateNode,
    multi_input_node: multiInputNode,
  },
  definition: {
    edges: {
      e1: {
        id: "e1",
        source: "n1",
        target: "n2",
        sourceHandle: "template",
        targetHandle: "template",
      },
      e2: {
        id: "e2",
        source: "n2",
        target: "n3",
        sourceHandle: "text",
        targetHandle: "prompt",
      },
    },
    nodes: {
      n1: {
        id: "n1",
        type: "simple_node",
        autorun: false,
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      n2: {
        id: "n2",
        type: "template_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
      n3: {
        id: "n3",
        type: "multi_input_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    },
  },
  state: {
    n1: {
      input: {},
      status: "idle",
    },
    n2: {
      input: {
        variables: [
          {
            id: "World",
            name: "World",
            description: "",
            type: "text",
            defaultValue: "mars",
          },
        ],
      },
      status: "idle",
    },
    n3: {
      input: {},
      status: "idle",
    },
  },
};

export const simpleLoop: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
  },
  definition: {
    edges: {
      e1: {
        id: "e1",
        source: "n1",
        target: "n1",
        sourceHandle: "template",
        targetHandle: "template",
      },
    },
    nodes: {
      n1: {
        id: "n1",
        type: "simple_node",
        editor: {
          position: { x: 0, y: 0 },
        },
      },
    },
  },
  state: {
    "1": {
      input: {},
      status: "idle",
    },
  },
};

export const missingStates: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
    template_node: templateNode,
    multi_input_node: multiInputNode,
    async_node: asyncNode,
  },
  definition: {
    nodes: {
      n1: {
        id: "n1",
        type: "simple_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      n2: {
        id: "n2",
        type: "template_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      n3: {
        id: "n3",
        type: "async_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },

      n4: {
        id: "n4",
        type: "multi_input_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
    },
    edges: {
      e1: {
        id: "e1",
        source: "n1",
        target: "n2",
        sourceHandle: "template",
        targetHandle: "template",
      },
      e2: {
        id: "e2",
        source: "n2",
        target: "n3",
        sourceHandle: "text",
        targetHandle: "prompt",
      },
      e3: {
        id: "e3",
        source: "n2",
        target: "n4",
        sourceHandle: "text",
        targetHandle: "prompt",
      },
      e4: {
        id: "e4",
        source: "n3",
        target: "n4",
        sourceHandle: "completion",
        targetHandle: "completion",
      },
    },
  },
  state: {
    n3: {
      status: "idle",
      input: { template: "This is a test!" },
    },
  },
};

export const multiInput: Chain = {
  nodeTypes: {
    simple_node: simpleNode,
    template_node: templateNode,
    multi_input_node: multiInputNode,
  },
  definition: {
    nodes: {
      n1: {
        id: "n1",
        type: "simple_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      n2: {
        id: "n2",
        type: "template_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
      n3: {
        id: "n3",
        type: "multi_input_node",
        autorun: true,
        editor: { position: { x: 0, y: 0 } },
      },
    },
    edges: {
      e1: {
        id: "e1",
        source: "n1",
        target: "n2",
        sourceHandle: "template",
        targetHandle: "template",
      },
      e2: {
        id: "e2",
        source: "n1",
        target: "n3",
        sourceHandle: "template",
        targetHandle: "prompt",
      },
      e3: {
        id: "e3",
        source: "n2",
        target: "n3",
        sourceHandle: "text",
        targetHandle: "completion",
      },
    },
  },
  state: {
    "1": {
      status: "idle",
      input: { template: "This is a test!" },
    },
  },
};
