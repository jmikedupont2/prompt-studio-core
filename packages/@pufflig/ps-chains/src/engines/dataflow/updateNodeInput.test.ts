import { simpleChain, singleNodeChain } from "../../mocks/chains";
import { updateNodeInput } from "./updateNodeInput";

test("updateNodeInput - singleNodeChain", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await updateNodeInput("1", { template: "Hello World" }, singleNodeChain, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(res).toMatchSnapshot();
});

test("updateNodeInput - simpleChain, the children nodes are not updated", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await updateNodeInput("n1", { template: "Hello World" }, simpleChain, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(res).toMatchSnapshot();
});
