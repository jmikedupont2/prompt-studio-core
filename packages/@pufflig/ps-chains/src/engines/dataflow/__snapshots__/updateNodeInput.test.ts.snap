// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateNodeInput - simpleChain, the children nodes are not updated 1`] = `
{
  "n1": {
    "input": {
      "template": "Hello World",
    },
    "status": "idle",
  },
  "n2": {
    "input": {
      "variables": [
        {
          "defaultValue": "mars",
          "description": "",
          "id": "World",
          "name": "World",
          "type": "text",
        },
      ],
    },
    "status": "idle",
  },
  "n3": {
    "input": {},
    "status": "idle",
  },
}
`;

exports[`updateNodeInput - singleNodeChain 1`] = `
{
  "1": {
    "input": {
      "template": "Hello World",
    },
    "status": "idle",
  },
}
`;
